govcd/nsxt_nat_rule.go

```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		return nil, fmt.Errorf("error creating NSX-T NAT rule: %s", err)
+	}
+
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("task failed while creating NSX-T NAT rule: %s", err)
+	}
+
+	// queryParameters (API side filtering) are not used because pretty much nothing is accepted as filter (such fields as
+	// name, description, ruleType and even ID are not allowed
+	allNatRules, err := egw.GetAllNatRules(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching all NAT rules: %s", err)
+	}
+
+	for index, singleRule := range allNatRules {
+		// Look for a matching rule
+		if singleRule.IsEqualTo(natRuleConfig) {
+			return allNatRules[index], nil
+
+		}
+	}
+	return nil, fmt.Errorf("rule '%s' of type '%s' not found after creation", natRuleConfig.Name, natRuleConfig.RuleType)
+}
+
+// Update allows users to update NSX-T NAT rule
+func (nsxtNat *NsxtNatRule) Update(natRuleConfig *types.NsxtNatRule) (*NsxtNatRule, error) {
+	client := nsxtNat.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules
+	apiVersion, err := client.checkOpenApiEndpointCompatibility(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	apiVersion = elevateNsxtNatRuleApiVersion(apiVersion, client)
+
+	if nsxtNat.NsxtNatRule.ID == "" {
+		return nil, fmt.Errorf("cannot update NSX-T NAT Rule without ID")
+	}
+
+	urlRef, err := nsxtNat.client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, nsxtNat.edgeGatewayId), nsxtNat.NsxtNatRule.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	returnObject := &NsxtNatRule{
+		NsxtNatRule:   &types.NsxtNatRule{},
+		client:        client,
+		edgeGatewayId: nsxtNat.edgeGatewayId,
+	}
+
+	err = client.OpenApiPutItem(apiVersion, urlRef, nil, natRuleConfig, returnObject.NsxtNatRule, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error updating NSX-T NAT Rule: %s", err)
+	}
+
+	return returnObject, nil
+}
+
+// Delete deletes NSX-T NAT rule
+func (nsxtNat *NsxtNatRule) Delete() error {
+	client := nsxtNat.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules
+	apiVersion, err := client.checkOpenApiEndpointCompatibility(endpoint)
+	if err != nil {
+		return err
+	}
+
+	apiVersion = elevateNsxtNatRuleApiVersion(apiVersion, client)
+
+	if nsxtNat.NsxtNatRule.ID == "" {
+		return fmt.Errorf("cannot delete NSX-T NAT rule without ID")
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, nsxtNat.edgeGatewayId), nsxtNat.NsxtNatRule.ID)
+	if err != nil {
+		return err
+	}
+
+	err = client.OpenApiDeleteItem(apiVersion, urlRef, nil, nil)
+
+	if err != nil {
+		return fmt.Errorf("error deleting NSX-T NAT Rule: %s", err)
+	}
+
+	return nil
+}
+
+// IsEqualTo allows to check if a rule has exactly the same fields (except ID) to the supplied rule
+// This validation is very tricky because minor version changes impact how fields are return.
+// This function relies on most common and stable fields:
```

Comment:
I understand that `Type` and `RuleType` may not both be there, but a comparison should be still viable. What am I missing?

Also, could you make a note on why these fields are not reliably comparable?

* InternalPort,
* FirewallMatch,
* Priority,      
* Version,

Reply:
>Also, could you make a note on why these fields are not reliably comparable?

Good question which led me to another round of thinking. I was carefully trying to match as many fields as possible (while still having comprehensible comparison code). Main problem for many fields is - behavior changes across API versions which can make this matching very messy - value comparisons differ based on version:

* `Type` and `RuleType` - the trick is - on API V36.0 one can set `RuleType` (deprecated) or `Type`  field value (except the new `REFLEXIVE` which explicitly requires `Type` field). After READ it would return both `RuleType` and `Type` being set with the same value, _unless_ it is `REFLEXIVE`(in that case it returns `ruleType=null` and `type=REFLEXIVE`)
* FirewallMatch - it exists only since API 35.2+ and has a default starting this version
* InternalPort - is deprecated since API V35.0+ and is replaced by DnatExternalPort
* Priority - is available only in API V35.2+
* Version - it is something that is automatically handled by API. When creating - you must specify none, but it sets version to 0. When updating one must specify the last version read, and again it will automatically increment this value after update. (probably it is meant to avoid concurrent updates)

We could probably even write smart rules for tested versions, but what I am more afraid of is that even using a known API version might break in future VCD versions as we already saw a few times before. I know that in future versions NAT should return ID after POST and that should not be required at all, but now we still have it.

For now I have put this comment as well into function docs.

Any other ideas? I will sleep on it to see if anything more pops up now as I refreshed this context.